File: src/main/java/pcb/units/concrete/time/Days.java
package pcb.units.concrete.time;

import pcb.units.amount.BigDecimalAmount;
import pcb.units.base.AbstractUnitAmount;
import pcb.units.base.UnitAmount;
import pcb.units.dimensions.time.Time;
import pcb.units.dimensions.time.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Days
		extends AbstractUnitAmount<BigDecimalAmount, TimeUnit>
		implements Time {

	// region constructors

	public Days(Number value) {
		this(value.toString());
	}

	public Days(String value) {
		this(new BigDecimal(value));
	}

	public Days(BigDecimal value) {
		this(new BigDecimalAmount(value));
	}

	public Days(BigDecimalAmount amount) {
		super(amount, DayUnit.instance);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Days plus(UnitAmount<BigDecimalAmount, TimeUnit> other, MathContext mathContext) {
		Days meters = new Days(other.getAmountIn(getUnit()));
		return new Days(getAmount().plus(meters.getAmount(), mathContext));
	}

	@Override
	public Days minus(UnitAmount<BigDecimalAmount, TimeUnit> other, MathContext mathContext) {
		Days meters = new Days(other.getAmountIn(getUnit()));
		return new Days(getAmount().minus(meters.getAmount(), mathContext));
	}

	@Override
	public Days multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Days(getAmount().multipliedBy(other, mathContext));
	}

	@Override
	public Days dividedBy(BigDecimal other, MathContext mathContext) {
		return new Days(getAmount().dividedBy(other, mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Days.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.DayUnit
import pcb.uwu.unit.base.TimeUnit

class Days : Time
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, DayUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, DayUnit)

    // region UnitAmount

    override fun plus(time: UnitAmount<TimeUnit>) =
        Days(this.amount + (time to DayUnit).amount)

    override fun minus(time: UnitAmount<TimeUnit>) =
        Days(this.amount - (time to DayUnit).amount)

    override fun times(scalar: Number) =
        Days(this.amount * scalar)

    override fun div(scalar: Number) =
        Days(this.amount / scalar)

    // endregion
}
